plugins/com.google.cloud.tools.eclipse.appengine.compat.cte13/src/com/google/cloud/tools/eclipse/appengine/compat/cte13/CloudToolsEclipseProjectNotifier.java

```diff
@@ -0,0 +1,151 @@
+import org.eclipse.core.runtime.SubMonitor;
+import org.eclipse.core.runtime.jobs.Job;
+import org.eclipse.jface.dialogs.MessageDialog;
+import org.eclipse.swt.widgets.Shell;
+import org.eclipse.ui.IStartup;
+import org.eclipse.ui.IWorkbench;
+import org.eclipse.ui.PlatformUI;
+
+/**
+ * Looks for projects using the per-library classpath container approach used in versions prior
+ * Cloud Tools for Eclipse 1.4, and notifies the user to upgrade.
+ */
+public class CloudToolsEclipseProjectNotifier implements IStartup {
+  private static final Logger logger =
+      Logger.getLogger(CloudToolsEclipseProjectNotifier.class.getName());
+  private IWorkbench workbench;
+  private IWorkspace workspace;
+
+  @Override
+  public void earlyStartup() {
+    workbench = PlatformUI.getWorkbench();
+    workspace = ResourcesPlugin.getWorkspace();
+
+    Job projectUpdater = new WorkspaceJob("Updating projects for Cloud Tools for Eclipse") {
+      @Override
+      public IStatus runInWorkspace(IProgressMonitor monitor) throws CoreException {
+        SubMonitor progress = SubMonitor.convert(monitor, 40);
+        progress.subTask("Searching for projects requiring updating");
+        Collection<IProject> projects = findCandidates(progress.newChild(10));
+        if (projects.isEmpty()) {
+          return Status.OK_STATUS;
+        }
+        projects = promptUser(projects, progress.newChild(5));
+        if (projects.isEmpty()) {
+          return Status.OK_STATUS;
+        }
+        progress.subTask("Updating projects");
+        return upgradeProjects(projects, progress.newChild(25));
+      }
+    };
+    projectUpdater.setRule(workspace.getRoot());
+    projectUpdater.setUser(true);
+    projectUpdater.schedule(500);
+  }
+
+  /**
+   * Scan the current projects to identify those requiring upgrading.
+   */
+  protected Collection<IProject> findCandidates(SubMonitor progress) {
+    List<IProject> projects = new ArrayList<>();
+
+    IProject[] allProjects = workspace.getRoot().getProjects();
+    progress.setWorkRemaining(allProjects.length);
+    for (IProject project : allProjects) {
+      if (CloudToolsEclipseProjectUpdater.hasOldContainers(project)) {
+        projects.add(project);
+      }
+      progress.worked(1);
+    }
+    Collections.sort(projects, Ordering.usingToString());
+    return projects;
+  }
+
+
+  /**
+   * Prompt the user to select the projects to upgrade.
+   */
+  protected Collection<IProject> promptUser(final Collection<IProject> projects,
+      SubMonitor progress) {
+    Preconditions.checkArgument(!projects.isEmpty(), "no projects specified!"); // $NON-NLS-1$
+    progress.setBlocked(StatusUtil.info(this, "Prompting user to upgrade"));
+    final boolean[] proceed = new boolean[1];
+    workbench.getDisplay().syncExec(new Runnable() {
+      public void run() {
+        StringBuilder sb = new StringBuilder(
+            "The following projects must be updated for Cloud Tools for Eclipse:\n");
```

Comment:
Ideally this would be in a message bundle.

Reply:
ok